### PR TITLE
Support StatefulSetAutoDeletePVC feature

### LINKED ISSUE #813: [feature request] StatefulSet PVC auto-delete implementation
**What would you like to be added**:
Kubernetes has merged PR https://github.com/kubernetes/kubernetes/pull/99728 . Advanced StatefulSet should cherry-pick it to keep compatible with upstream.

### PR DESCRIPTION
**Pull request by veophi**
### Ⅰ. Describe what this PR does
Cherry pick from [kubernetes/kubernetes#99728](https://github.com/kubernetes/kubernetes/pull/99728).

### Ⅱ. Does this pull request fix one issue?

fixes #813

### CODE REVIEW

File: test/e2e/framework/pv_util.go [hunk 1]
```diff
@@ -39,3 +61,49 @@ func createPV(c clientset.Interface, pv *v1.PersistentVolume) (*v1.PersistentVol
...
+	for _, sc := range list.Items {
+		if isDefaultAnnotation(sc.ObjectMeta) {
+			if len(scName) != 0 {
+				return "", fmt.Errorf("Multiple default storage classes found: %q and %q", scName, sc.Name)
+			}
+			scName = sc.Name
+		}
+	}
+	if len(scName) == 0 {
+		return "", fmt.Errorf("No default storage class found")
```
[the remark this answers is not in the document]
Comment (veophi) [reply test/e2e/framework/pv_util.go:79]: ignore

File: pkg/controller/statefulset/stateful_set_control_test.go [hunk 2]
```diff
@@ -2686,90 +2944,74 @@ func (spc *fakeStatefulPodControl) setPodReady(set *appsv1beta1.StatefulSet, ord
 	condition := v1.PodCondition{Type: v1.PodReady, Status: v1.ConditionTrue}
 	podutil.UpdatePodCondition(&pod.Status, &condition)
 	fakeResourceVersion(pod)
-	spc.podsIndexer.Update(pod)
-	return spc.podsLister.Pods(set.Namespace).List(selector)
+	om.podsIndexer.Update(pod)
+	return om.podsLister.Pods(set.Namespace).List(selector)
+}
+
+func (om *fakeObjectManager) setPodAvailable(set *appsv1beta1.StatefulSet, ordinal int, lastTransitionTime time.Time) ([]*v1.Pod, error) {
```
[the remark this answers is not in the document]
Comment (veophi) [reply pkg/controller/statefulset/stateful_set_control_test.go:2951]: ignore

File: pkg/controller/statefulset/stateful_pod_control.go [hunk 3]
```diff
@@ -137,15 +201,92 @@ func (spc *realStatefulPodControl) UpdateStatefulPod(set *appsv1beta1.StatefulSe
...
+			klog.V(4).Infof("Expected claim %s missing, continuing to pick up in next iteration.")
+		case err != nil:
+			return fmt.Errorf("Could not retrieve claim %s not found for %s when checking PVC deletion policy: %w", claimName, pod.Name, err)
+		default:
+			if !claimOwnerMatchesSetAndPod(claim, set, pod) {
+				needsUpdate := updateClaimOwnerRefForSetAndPod(claim, set, pod)
+				if needsUpdate {
+					err := spc.objectMgr.UpdateClaim(claim)
+					if err != nil {
+						return fmt.Errorf("Could not update claim %s for delete policy ownerRefs: %w", claimName, err)
```
[the remark this answers is not in the document]
Comment (veophi) [reply pkg/controller/statefulset/stateful_pod_control.go:251]: ignore

**Review by FillZpp — APPROVED** [review]
/lgtm

**Pull request merged**